src/webgpu/shader/validation/shader_validation_test.ts

```diff
@@ -55,6 +55,55 @@ export class ShaderValidationTest extends GPUTest {
     });
   }
 
+  /**
+   * Add a test expectation for whether a createShaderModule call issues a warning.
+   *
+   * @example
+   * ```ts
+   * t.expectCompileWarning(true, `wgsl code`); // Expect compile success and any warning message
+   * t.expectCompileWarning(false, `wgsl code`); // Expect compile success and no warning messages
+   * ```
+   */
+  expectCompileWarning(expectWarning: boolean, code: string) {
+    let shaderModule: GPUShaderModule;
+    this.expectGPUError(
+      'validation',
+      () => {
+        shaderModule = this.device.createShaderModule({ code });
+      },
+      false
+    );
+
+    const error = new ErrorWithExtra('', () => ({ shaderModule }));
+    this.eventualAsyncExpectation(async () => {
+      const compilationInfo = await shaderModule!.getCompilationInfo();
+
+      // MAINTENANCE_TODO: Pretty-print error messages with source context.
+      const messagesLog = compilationInfo.messages
+        .map(m => `${m.lineNum}:${m.linePos}: ${m.type}: ${m.message}`)
+        .join('\n');
+      error.extra.compilationInfo = compilationInfo;
+
+      if (compilationInfo.messages.some(m => m.type === 'warning')) {
+        if (expectWarning === true) {
```

Comment:
nit:  Comparing a bool to true can always be eliminated.  I'd just rewrite as   `if (expectWarning)`  ?

coders choice though.